File: frontend/src/api/connectedAgents.ts
/**
 * Task M3c-2 — frontend API client for the per-user "connected agents"
 * surface (MCP clients that have completed the DCR + browser-OAuth consent
 * flow and hold an active grant).
 *
 * Backend contract (M3c, core):
 *   - GET    /api/users/me/connected-agents          → ConnectedAgent[]
 *   - DELETE /api/users/me/connected-agents/{client}  → { revoked: number }
 *
 * The shared axios client (`./client`) prefixes `/api` and attaches the JWT
 * bearer, so both calls are authenticated as the logged-in user. Mirrors the
 * hand-written idiom of `agents.ts` / `oauthConsent.ts`: the backend Pydantic
 * boundary is the single source of truth — adding a field here without a
 * matching backend change is a contract violation.
 */
import api from './client';

/** One MCP client the user has granted access to. */
export interface ConnectedAgent {
  client_id: string;
  client_name: string;
  scopes: string[];
  granted_at: string;
}

/** Shape returned by DELETE — number of grants revoked for the client. */
export interface RevokeResult {
  revoked: number;
}

export const connectedAgentsApi = {
  async list(): Promise<ConnectedAgent[]> {
    const { data } = await api.get<ConnectedAgent[]>(
      '/users/me/connected-agents',
    );
    return data;
  },
  async revoke(clientId: string): Promise<RevokeResult> {
    const { data } = await api.delete<RevokeResult>(
      `/users/me/connected-agents/${clientId}`,
    );
    return data;
  },
};

File: backend/app/schemas/connected_agents.py
"""Wire shapes for the user-facing connected-agents surface (M3b-2).

Back the two bearer-authed endpoints in ``app/api/users.py`` that let a user
see and revoke the external OAuth clients (MCP agents) holding an active
refresh-token grant for them:

  * ``GET    /users/me/connected-agents``             -> ``list[ConnectedAgent]``
  * ``DELETE /users/me/connected-agents/{client_id}`` -> ``RevokeAgentResponse``

Per AGENTS.md § jvspatial Object-Spatial Contract, request/response shapes live
here, never inline in the handler module.
"""

from __future__ import annotations

from datetime import datetime
from typing import List

from pydantic import BaseModel, Field


class ConnectedAgent(BaseModel):
    """One distinct OAuth client holding an active grant for the caller.

    Multiple active refresh tokens for the same client (grant-family
    rotations) collapse into a single row; ``granted_at`` is the most-recent
    token's ``created_at``.
    """

    client_id: str = Field(..., description="The OAuth client's public id")
    client_name: str = Field(
        default="",
        description="Human-readable client name (falls back to client_id)",
    )
    scopes: List[str] = Field(
        default_factory=list,
        description="Granted scopes for this client's active grant",
    )
    granted_at: datetime = Field(
        ...,
        description="Most-recent active-token created_at for this client",
    )


class RevokeAgentResponse(BaseModel):
    """Result of revoking a client's grant for the caller.

    ``revoked`` is the number of the caller's active refresh tokens for the
    client that were deactivated by the request.
    """

    revoked: int = Field(
        ..., description="Count of the caller's active tokens deactivated"
    )


__all__ = [
    "ConnectedAgent",
    "RevokeAgentResponse",
]
